ipa-core/src/protocol/ipa_prf/quicksort.rs

```diff
@@ -54,65 +53,96 @@ where
     for<'a> &'a AdditiveShare<K>: IntoIterator<Item = AdditiveShare<K::Element>>,
     K: SharedValue + Field + CustomArray<Element = Boolean>,
 {
-    // expected amount of recursion: compute ceil(log_2(list.len()))
-    let expected_depth = std::mem::size_of::<usize>() * 8 - list.len().leading_zeros() as usize;
-    // create stack
-    let mut stack: Vec<(C, usize, usize)> = Vec::with_capacity(expected_depth);
+    let mut ranges_for_next_pass = Vec::with_capacity(ranges_to_sort.len() * 2);
+    let mut quicksort_pass = 1;
+
+    // iterate through all of the potentially incorrectly ordered ranges
+    // make one pass, comparing each element to the pivot and splitting into two more
+    // potentially incorrectly ordered ranges
+    while !ranges_to_sort.is_empty() {
+        // compute the total number of comparisons that will be needed
+        let mut num_comparisons_needed = 0;
+        for range in &ranges_to_sort {
+            if range.len() > 1 {
+                num_comparisons_needed += range.len() - 1;
+            }
+        }
+
+        let c = ctx
+            .narrow(&BitOpStep::from(quicksort_pass))
+            .set_total_records(num_comparisons_needed);
+        let cmp_ctx = c.narrow(&Step::Compare);
+        let rvl_ctx = c.narrow(&Step::Reveal);
 
-    // initialize stack
-    stack.push((ctx, 0usize, list.len()));
+        let mut futures = Vec::with_capacity(num_comparisons_needed);
+        let mut counter = 0;
 
-    // iterate through quicksort recursions
-    while let Some((ctx, b_l, b_r)) = stack.pop() {
-        // start of quicksort function
         // check whether sort is needed
-        if b_l + 1 < b_r {
+        for range in &ranges_to_sort {
+            if range.len() <= 1 {
+                continue;
+            }
             // set up iterator
-            let mut iterator = list[b_l..b_r].iter().map(get_key);
+            let mut iterator = list[range.clone()].iter().map(get_key);
             // first element is pivot, apply key extraction function f
```

Comment:
`range` is a reference. I had to do this because removing the call to clone says:

```
the type `[S]` cannot be indexed by `&std::ops::Range<usize>`
```

This made me sad, but a `Range` is just a struct made of two `usize`, so I think it's a fairly cheap clone.